src/raft.rs

```diff
@@ -2096,176 +2151,154 @@ impl<T: Storage> Raft<T> {
-        self.promotable
-    }
+        self.pending_request_snapshot = INVALID_INDEX;
 
-    /// # Errors
-    ///
-    /// * `id` is already a voter.
-    /// * `id` is already a learner.
-    fn add_voter_or_learner(&mut self, id: u64, learner: bool) -> Result<()> {
-        debug!(
+        info!(
             self.logger,
-            "adding node (learner: {learner}) with ID {id} to peers.",
-            learner = learner,
-            id = id,
+            "restored snapshot";
+            "commit" => self.raft_log.committed,
+            "last_index" => self.raft_log.last_index(),
+            "last_term" => self.raft_log.last_term(),
+            "snapshot_index" => snap_index,
+            "snapshot_term" => snap_term,
         );
 
-        let result = if learner {
-            let progress = Progress::new(self.raft_log.last_index() + 1, self.max_inflight);
-            self.mut_prs().insert_learner(id, progress)
-        } else if self.prs().learner_ids().contains(id) {
-            self.mut_prs().promote_learner(id)
-        } else {
-            let progress = Progress::new(self.raft_log.last_index() + 1, self.max_inflight);
-            self.mut_prs().insert_voter(id, progress)
-        };
-
-        if let Err(e) = result {
-            error!(self.logger, ""; "e" => %e);
-            return Err(e);
-        }
-        if self.id == id {
-            self.promotable = !learner;
-        };
-        // When a node is first added/promoted, we should mark it as recently active.
-        // Otherwise, check_quorum may cause us to step down if it is invoked
-        // before the added node has a chance to communicate with us.
-        self.mut_prs().get_mut(id).unwrap().recent_active = true;
-        result
-    }
-
-    /// Adds a new node to the cluster.
-    ///
-    /// # Errors
-    ///
-    /// * `id` is already a voter.
-    /// * `id` is already a learner.
-    pub fn add_node(&mut self, id: u64) -> Result<()> {
-        self.add_voter_or_learner(id, false)
-    }
-
-    /// Adds a learner node.
-    ///
-    /// # Errors
-    ///
-    /// * `id` is already a voter.
-    /// * `id` is already a learner.
-    pub fn add_learner(&mut self, id: u64) -> Result<()> {
-        self.add_voter_or_learner(id, true)
+        true
     }
 
-    /// Removes a node from the raft.
-    ///
-    /// # Errors
-    ///
-    /// * `id` is not a voter or learner.
-    pub fn remove_node(&mut self, id: u64) -> Result<()> {
-        self.mut_prs().remove(id)?;
-
-        // do not try to commit or abort transferring if there are no voters in the cluster.
-        if self.prs().voter_ids().is_empty() {
-            return Ok(());
+    /// Updates the in-memory state and, when necessary, carries out additional actions
+    /// such as reacting to the removal of nodes or changed quorum requirements.
+    pub fn post_conf_change(&mut self) -> ConfState {
+        info!(self.logger, "switched to configuration"; "config" => ?self.prs.conf());
```

Comment:
In fact, it was already switched during `restore`, a little confused about the log and the function command.

Reply:
Indeed, that's why the logs say "switched".